tensorboard/plugins/debugger/debugger_plugin_test.py

```diff
@@ -0,0 +1,436 @@
+    writer.Close()
+
+    # Populate the log directory with debugger event for run 'run_foo'.
+    run_foo_directory = os.path.join(self.log_dir, 'run_foo')
+    os.mkdir(run_foo_directory)
+    file_prefix = tf.compat.as_bytes(
+        os.path.join(run_foo_directory, 'events.debugger'))
+    writer = tf.pywrap_tensorflow.EventsWriter(file_prefix)
+    writer.WriteEvent(
+        self._CreateEventWithDebugNumericSummary(
+            device_name=device_name,
+            op_name='layers/Variable',
+            output_slot=0,
+            wall_time=4242,
+            step=42,
+            list_of_values=(list(range(12)) +
+                            [float(tf.int16.as_datatype_enum), 1.0, 8.0])))
+    writer.Close()
+
+    # Start a server that will receive requests and respond with health pills.
+    self.multiplexer = event_multiplexer.EventMultiplexer({
+        '.': self.log_dir,
+        'run_foo': run_foo_directory,
+    })
+    self.debugger_data_server_grpc_port = 4242
+
+    # Fake threading behavior so that threads are synchronous.
+    class FakeThread(object):
+
+      def __init__(self, target):
+        """A 'thread' that just synchronously calls a method.
+
+        Args:
+          target: The function to call.
+        """
+        self._target = target
+
+      def start(self):  # pylint: disable=invalid-name
+        """Starts the thread."""
+        self._target()
+
+    mock.patch('threading.Thread', FakeThread).start()
+
+    self.mock_debugger_data_server = mock.create_autospec(
+        debugger_server_lib.DebuggerDataServer)
+    self.mock_debugger_data_server_class = mock.create_autospec(
+        debugger_server_lib.DebuggerDataServer,
+        return_value=self.mock_debugger_data_server)
+    mock.patch('tensorboard.plugins.debugger.'
+               'debugger_server_lib.DebuggerDataServer',
+               self.mock_debugger_data_server_class).start()
+
+    self.context = base_plugin.TBContext(
+        logdir=self.log_dir, multiplexer=self.multiplexer)
+    self.plugin = debugger_plugin.DebuggerPlugin(
+        self.context,
+        debugger_data_server_grpc_port=self.debugger_data_server_grpc_port)
+    wsgi_app = application.TensorBoardWSGIApp(
+        self.log_dir, [self.plugin], self.multiplexer, reload_interval=0)
+    self.server = werkzeug_test.Client(wsgi_app, wrappers.BaseResponse)
+
+    # The debugger data server should be started at the correct port.
+    self.mock_debugger_data_server_class.assert_called_once_with(
+        self.debugger_data_server_grpc_port, self.log_dir)
+
+    mock_debugger_data_server = self.mock_debugger_data_server
+    start = mock_debugger_data_server.start_the_debugger_data_receiving_server
+    start.assert_called_once()
+
+  def tearDown(self):
+    # Remove the directory with debugger-related events files.
+    shutil.rmtree(self.log_dir, ignore_errors=True)
+    mock.patch.stopall()
+
+  def _CreateEventWithDebugNumericSummary(
```

Comment:
@chihuahua If I understand correctly, the debugger plugin is different from other plugins like histograms in that: 1) it writes summary files, but 2) the user doesn't write events explicitly. Instead, the plugin receives data from TF runtimes and store them in summary files. Do you think there is room for improvements in the way the summary data is created and tested here?